src/main/src/main/java/org/geoserver/security/decorators/DefaultSecureCatalogFactory.java

```diff
@@ -29,32 +37,145 @@ public boolean canSecure(Class clazz) {
         else if (CoverageStoreInfo.class.isAssignableFrom(clazz))
-            return new SecuredCoverageStoreInfo((CoverageStoreInfo) object, policy);
+            return new SecuredCoverageStoreInfo(unwrap((CoverageStoreInfo) object), policy);
         else if (DataStoreInfo.class.isAssignableFrom(clazz))
-            return new SecuredDataStoreInfo((DataStoreInfo) object, policy);
+            return new SecuredDataStoreInfo(unwrap((DataStoreInfo) object), policy);
         else if (FeatureTypeInfo.class.isAssignableFrom(clazz))
-            return new SecuredFeatureTypeInfo((FeatureTypeInfo) object, policy);
+            return new SecuredFeatureTypeInfo(unwrap((FeatureTypeInfo) object), policy);
         else if (LayerInfo.class.isAssignableFrom(clazz))
-            return new SecuredLayerInfo((LayerInfo) object, policy);
+            return new SecuredLayerInfo(unwrap((LayerInfo) object), policy);
+        else if (WMTSLayerInfo.class.isAssignableFrom(clazz))
+            return new SecuredWMTSLayerInfo(unwrap((WMTSLayerInfo) object), policy);
         else
-            throw new IllegalArgumentException("Don't know how to wrap");
+            throw new IllegalArgumentException("Don't know how to wrap " + object);
     }
-
     /**
      * Returns {@link ExtensionPriority#LOWEST} since the wrappers generated by
      * this factory
+     * @return {@link ExtensionPriority#LOWEST}
      */
+    @Override
     public int getPriority() {
         return ExtensionPriority.LOWEST;
     }
 
+    private void logDoubleWrap(Object unwrapped, Object orig) {
+        String msg = String.format("Tried to double secure: %s already securing %s", orig, unwrapped);
+        LOGGER.warning(msg);
+    }
+
+    private void logDoubleWrap(final DataStoreInfo store) {
+        int proxies = 0;
+        int secured = 0;
+        DataStoreInfo object = store;
+        while (true) {
+            DataStoreInfo unwrapped;
+            if (object instanceof SecuredDataStoreInfo) {
+                secured++;
+                unwrapped = ((SecuredDataStoreInfo) object).unwrap(DataStoreInfo.class);
+            } else {
+                unwrapped = ModificationProxy.unwrap(object);
+                boolean isProxy = object != unwrapped;
+                if (isProxy) {
+                    proxies++;
+                } else {
+                    break;
+                }
+            }
+            object = unwrapped;
+        }
+
+        String msg = String.format("Double securing %s: proxies: %,d, SecuredDataStoreInfos: %,d",
+                store.getName(), proxies, secured);
+        LOGGER.warning(msg);
+    }
+
+    private WMTSLayerInfo unwrap(WMTSLayerInfo object) {
+        WMTSLayerInfo unwrapped = ModificationProxy.unwrap(object);
+        boolean isProxy = object != unwrapped;
+        if(unwrapped instanceof SecuredWMTSLayerInfo) {
+            logDoubleWrap(unwrapped, object);
```

Comment:
This confused me for a bit but if I understand what's going on this is to ensure it's not wrapped by a security proxy before wrapping it with one.  So the warning about being double wrapped is that this is about to be wrapped, not the combination of modification and security wrappers?

Reply:
The warning is more to indicate that somewhere, a secured Object is being doubly secured. I have not seen this for some of the Info Object types handled in this class, but it has been observed for at least DataStoreInfo objects. In theory, we shouldn't see this log generated, as I would think we wouldn't want to try to secure an already secured Object.